### ISSUE TITLE
parameter type declaration does not work with variables

### ISSUE THREAD

**Issue by flypenguin** [hashicorp/terraform#9578]
(ZIP: "...", line ..." refers to a file and line number in the ZIP archive for reference)
- a module exports a string value (in this example: a aws_security_group._.id) as `module.thing.thang` (ZIP: `modules/vpc-baselayout/main.tf`, line 34)
- another module defines a parameter type explicitly as string (ZIP: `modules/pet-host/main.tf`, line 5)
- using `thong = ["${module.thing.thang}"]` as input for the 2nd module results in an error ("`...should be a string`" (ZIP: `vpc.tf`, line 28), which is IMHO just wrong. Also, _not_ using a variable for the 2nd module's parameter _does_ work (`thong = ["wo-hooo", "bo-yah"]` for example)
- if you change how the module's parameter is used (see ZIP: `modules/pet-host/main.tf`, line 28) then it works again, but in my understanding you are now passing a list with lists. Which is IMHO also plainly wrong, but definitively confusing.

A complete ready-to-run example can be [downloaded from dropbox](https://www.dropbox.com/s/oa8d3r6a86fp1mz/bugreport-terraform-listparams.zip?dl=0).
### Terraform Version

0.7.6, 0.7.7
### Affected Resource(s)

I guess all, but at least aws_autoscaling_group, aws_instance
### Terraform Configuration Files

[download ZIP from dropbox](https://www.dropbox.com/s/oa8d3r6a86fp1mz/bugreport-terraform-listparams.zip?dl=0).
### Debug Output

A `DEBUG_OUTPUT.txt` is included in the ZIP file (`TF_LOG=DEBUG terraform plan`)
### Panic Output

N/A
### Expected Behavior

It should work.
### Actual Behavior

An error message

```
There are warnings and/or errors related to your configuration. Please
fix these before continuing.

Errors:

  * aws_instance.host: vpc_security_group_ids: should be a list
```
### Steps to Reproduce
1. `terraform plan`

**Comment by thefirstofthe300** [comment]
I just ran into what I believe to be this same issue.

```
terraform version
Terraform v0.11.14
+ provider.google v2.10.0
```

Essentially, it seems that computed strings passed into lists used as variables result in the list being evaluated by the module as empty strings (or maps...but whatever it is, it's the wrong type). My code does essentially the same thing that @flypenguin's does; however, this bug also triggers if a data source is used.

Do we know if this issue still exists in TF v0.12? It seems to have stuck around from some 7 minor versions and still hasn't been fixed.

**Comment by teamterraform** [comment]
Hi all!

Indeed, in Terraform 0.11 and earlier there was a limitation where a list containing any unknown value became an unknown value itself, and in the process Terraform would lose the type information for it, so it would produce this confusing error.

This problem has been fixed in two ways in Terraform 0.12: firstly, Terraform is now able to retain lists that have a mixture of known and unknown elements, and secondly Terraform is able to track type information even when a value is unknown, so Terraform would still know that the result is a list of strings even if the list itself or any of its elements are unknown.

This was a common glitch with Terraform 0.11 and earlier and so we didn't notice this particular report of it when we were cleaning up after 0.12, so sorry we left this one sitting here without an update. Since Terraform 0.12 contains the fixes for this, we're going to close it out now. Thanks for reporting it!
